crates/wasi/src/preview2/host/tcp.rs

```diff
@@ -1,37 +1,100 @@
 };
-use crate::preview2::{Pollable, SocketResult, WasiView};
-use cap_net_ext::Blocking;
-use io_lifetimes::AsSocketlike;
-use rustix::io::Errno;
-use rustix::net::sockopt;
+use std::io;
 use std::net::SocketAddr;
+use std::task::{Context, Poll};
 use std::time::Duration;
-use tokio::io::Interest;
 use wasmtime::component::Resource;
 
-impl<T: WasiView> tcp::Host for T {}
+impl<T: WasiView> crate::preview2::bindings::sockets::tcp::Host for T {}
+
+/// The state of the TCP socket wrapper.
+///
+/// This represents the various states a socket can be in during the
+/// activities of binding, listening, accepting, and connecting.
+enum TcpState {
+    /// The initial state for a newly-created socket.
+    Default,
+
+    /// Binding started via `start_bind`.
+    BindStarted,
+
+    /// Binding finished via `finish_bind`. The socket has an address but
+    /// is not yet listening for connections.
+    Bound,
+
+    /// Listening started via `listen_start`.
+    ListenStarted,
+
+    /// The socket is now listening and waiting for an incoming connection.
+    Listening {
+        pending_result: Option<io::Result<(Box<dyn TcpSocket>, TcpReader, TcpWriter)>>,
+    },
+
+    /// An outgoing connection is started via `start_connect`.
+    Connecting {
+        future: Preview2Future<io::Result<(TcpReader, TcpWriter)>>,
+    },
+
+    /// An outgoing connection has been established.
+    Connected,
+}
+
+/// A `wasi:sockets/tcp::tcp-socket` instance.
+/// This is mostly glue code translating between WASI types and concepts (Tables,
+/// Resources, Pollables, ...) to their idiomatic Rust equivalents.
+pub struct TcpSocketWrapper {
+    inner: Box<dyn TcpSocket>,
+    tcp_state: TcpState,
+}
+
+impl TcpSocketWrapper {
+    fn new_input_stream(reader: TcpReader) -> InputStream {
+        InputStream::Host(Box::new(AsyncReadStream::new(reader)))
+    }
+
+    fn new_output_stream(writer: TcpWriter) -> OutputStream {
+        const SOCKET_READY_SIZE: usize = 1024 * 1024 * 1024;
 
-impl<T: WasiView> crate::preview2::host::tcp::tcp::HostTcpSocket for T {
+        Box::new(AsyncWriteStream::new(SOCKET_READY_SIZE, writer))
+    }
+}
+
+impl<T: WasiView> crate::preview2::bindings::sockets::tcp_create_socket::Host for T {
+    fn create_tcp_socket(
+        &mut self,
+        address_family: IpAddressFamily,
+    ) -> SocketResult<Resource<TcpSocketWrapper>> {
+        let socket = self
+            .ctx_mut()
+            .network
+            .new_tcp_socket(address_family.into())?;
+        let wrapper = TcpSocketWrapper {
+            inner: socket,
+            tcp_state: TcpState::Default,
+        };
+        let socket = self.table_mut().push(wrapper)?;
+        Ok(socket)
+    }
+}
+
+impl<T: WasiView> crate::preview2::bindings::sockets::tcp::HostTcpSocket for T {
     fn start_bind(
         &mut self,
-        this: Resource<tcp::TcpSocket>,
+        this: Resource<TcpSocketWrapper>,
```

Comment:
We use the term `Handle` for ip name lookup resources. We should probably settle on either `Wrapper` or `Handle` and use it for both.

Reply:
Good point. "Wrapper" would be a bit odd for `Network(Handle)` as that doesn;t wrap anything. So I went for a third option: `*Resource`, because of the 1:1 mapping to their WIT equivalents.